biomage/stage/stage.py

```diff
@@ -0,0 +1,297 @@
+
+    # Open each template and iterate through the documents. If we
+    # find a `fluxcd.io/automated` annotation, set it to the appropriate
+    # value depending on the pinning request.
+    for name, template in templates.items():
+        documents = yaml.load_all(template.text, Loader=yaml.SafeLoader)
+
+        for document in documents:
+            # disable automatic image fetching if pinning is on
+            if recursive_get(
+                document, "metadata", "annotations", "fluxcd.io/automated"
+            ):
+                document["metadata"]["annotations"]["fluxcd.io/automated"] = str(
+                    name not in pins
+                ).lower()
+
+            # pin chart version if pinning is on
+            if (
+                recursive_get(document, "spec", "chart", "ref")
+                and document["spec"]["chart"]["ref"] == "STAGING_CHART_REF"
+            ):
+                if name in pins:
+                    document["spec"]["chart"]["ref"] = get_latest_master_sha(
+                        document["spec"]["chart"], token
+                    )
+                else:
+                    document["spec"]["chart"]["ref"] = "master"
+
+            manifests.append(document)
+
+    manifests = yaml.dump_all(manifests)
+
+    # Generate a sandbox name and ask the user what they want theirs to be called.
+    sandbox_id = hashlib.md5(manifests.encode()).digest()
+    sandbox_id = anybase32.encode(sandbox_id, anybase32.ZBASE32).decode()
+
+    # Ask the user to provide one if they want
+    click.echo()
+    click.echo(click.style("Give a sandbox ID.", fg="yellow", bold=True))
+    click.echo(
+        "The sandbox ID must be no more than 26 characters long, consist of "
+        "lower case alphanumeric characters, or `-`, and must\n"
+        "start and end with an alphanumeric character. A unique ID generated from "
+        "the contents of the deployments and your pinning\n"
+        "choices has been provided as a default."
+    )
+    questions = [
+        {
+            "type": "input",
+            "name": "sandbox_id",
+            "message": "Provide an ID:",
+            "default": sandbox_id,
+            "validate": lambda i: SANDBOX_NAME_REGEX.match(i) and len(i) <= 26,
+        }
+    ]
+
+    click.echo()
+    sandbox_id = prompt(questions)
+    try:
+        sandbox_id = sandbox_id["sandbox_id"]
+    except Exception:
+        exit(1)
+
+    # Write sandbox ID
+    manifests = manifests.replace("STAGING_SANDBOX_ID", sandbox_id)
+
+    return manifests, sandbox_id
+
+
+@click.command()
+@click.argument("refs", nargs=-1)
+@click.option(
+    "--token",
+    "-t",
+    envvar="GITHUB_API_TOKEN",
+    required=True,
+    help="A GitHub Personal Access Token with the required permissions.",
+)
+@click.option(
+    "--org",
+    envvar="GITHUB_BIOMAGE_ORG",
+    default="biomage-ltd",
+    help="The GitHub organization to perform the operation in.",
+)
+def stage(token, org, refs):
```

Comment:
what does `refs` stand for? Is it actually pinned deployments?

Reply:
`refs` are the repositories and their versions that need to be deployed in a staging sandbox. For example, `api` is a ref, and by default corresponds to the repository `api` on `master`. `ui/5` is a ref, and it corresponds to the repository `ui` on the branch for pull request `5`.

Reply:
I see, I got confused then!